File: matmul/solution_omp_openacc/openmp-mm.f90
program matmul

  implicit none
  integer, parameter :: dp = selected_real_kind(14)
  integer, parameter :: nra=1500
  integer, parameter :: nca=2000
  integer, parameter :: ncb=1000
  integer :: i,j,k
  real(dp) :: a(nra,nca), b(nca,ncb), c(nra,ncb),sum
  integer, dimension(8) :: value
  real(dp) :: init_time, start_time, end_time, flops

  sum = 0d0
  flops = float(nra) * float(nca) * float(ncb)

  !  Start recording matrix initialization time (Init)
  call date_and_time(VALUES=value)
  init_time = float(value(6)*60) + float(value(7)) + float(value(8))/1000d0

  ! Initialize the multiplicand matrix
  do i = 1, nra
     do j = 1, nca
        a(i,j) = float(i+j)
     end do
  end do

  ! Initialize the multiplier matrix
  do i = 1, nca
     do j = 1, ncb
        b(i,j) = float(i*j)
     end do
  end do

  ! Start recording matrix multiplication (computation) time (Calc)
  call date_and_time(VALUES=value)
  start_time = float(value(6)*60) + float(value(7)) + float(value(8))/1000d0
  
  ! Create OpenMP threads and multiply matrix A and B
  !$omp parallel do private(sum) shared(a,b,c)
  do j = 1, nca
     do k = 1, ncb
        sum = 0d0
        do i = 1, nra
           sum = sum + a(i,j) * b(j,k)
        end do
        c(i,k) = sum
     end do
  end do
  !$omp end parallel do

  ! Stop recording computation time
  call date_and_time(VALUES=value)
  end_time = float(value(6)*60) + float(value(7)) + float(value(8))/1000d0

  ! Print timing results
  print '(a,f6.3,a,f6.3,a,f6.3)', 'OPENMP-MM: Init time: ',start_time - init_time, &
       '   Calc time: ', end_time - start_time, &
       '   GFlop/s: ', 1e-9*2.d0*flops/(end_time - start_time)

end program matmul
  

File: matmul/solution_omp_openacc/openmp-mm.c

/******************************************************************************
* FILE: openmp-mm.c
*
* DESCRIPTION:  
*   In this example code, the matrix multiplication operation is explicitly
*   parallelized by specifying OpenMP thread construct on main computation loop.
*   The number of threads depends on OMP_NUM_THREADS environment variable.
*
******************************************************************************/

#include <stdio.h>
#include <stdlib.h>

#define NRA 1500             /* Number of rows in matrix A */
#define NCA 2000             /* Number of columns (rows) in matrix A (B) */
#define NCB 1000             /* number of columns in matrix B */

#define dt(start, end)  ((end.tv_sec - start.tv_sec) + \
		      1/1000000.0*(end.tv_usec - start.tv_usec))

double a[NRA][NCA],          /* A - Multiplicand matrix */
       b[NCA][NCB],          /* B - Multiplier matrix */
       c[NRA][NCB];          /* C - Prodcut matrix */

main(){

	int i,j,k;
	struct timeval srun, scalc, ecalc;
	
	/* Start recording matrix initialization time (Init) */
	gettimeofday(&srun, NULL);

	/* Initialize the multiplicand matrix */
	for (i=0; i<NRA; i++) {
		for (j=0; j<NCA; j++) {
			a[i][j] = (double)(i+j);
		}
	}

	/* Initialize the multiplier matrix */
	for (i=0; i<NCA; i++) {
		for (j=0; j<NCB; j++) {
			b[i][j] = (double)(i*j);
		}
	}

	/* Start recording matrix multiplication (computation) time (Calc) */
	gettimeofday(&scalc, NULL);


	/* Create OpenMP threads and multiply matrix A and B */
	float sum;
#pragma omp parallel for private(sum,i,k,j)
	for (i=0; i<NRA; i++) {
		for (k=0; k<NCB; k++) {
			sum = 0.0;
			for (j=0; j<NCA; j++) {
				sum = sum + a[i][j] * b[j][k];
			}
			c[i][k] = sum;
		}
	}

	/* Stop recording computation time */
	gettimeofday(&ecalc, NULL);

	/* Print timing results */
	printf("OPENMP-MM: Init time: %3.02f   Calc time: %3.02f   GFlop/s: %3.02f\n",
		dt(srun, scalc), dt(scalc, ecalc), 1e-9*2.0*NRA*NCA*NCB / dt(scalc, ecalc));

	exit(0);
}
